// @unhead/schema-org/dist/chunks/index12.mjs
import { d as defineSchemaOrgResolver, ad as resolveWithBase, e as resolveRelation, a5 as imageResolver } from '../shared/schema-org.Cp6bpwL2.mjs';

const howToStepDirectionResolver = defineSchemaOrgResolver({
  cast(node) {
    if (typeof node === "string") {
      return {
        text: node
      };
    }
    return node;
  },
  defaults: {
    "@type": "HowToDirection"
  }
});

const howToStepResolver = defineSchemaOrgResolver({
  cast(node) {
    if (typeof node === "string") {
      return {
        text: node
      };
    }
    return node;
  },
  defaults: {
    "@type": "HowToStep"
  },
  resolve(step, ctx) {
    if (step.url)
      step.url = resolveWithBase(ctx.meta.url, step.url);
    if (step.image) {
      step.image = resolveRelation(step.image, ctx, imageResolver, {
        root: true
      });
    }
    if (step.itemListElement)
      step.itemListElement = resolveRelation(step.itemListElement, ctx, howToStepDirectionResolver);
    return step;
  }
});

const index = {
  __proto__: null,
  howToStepResolver: howToStepResolver
};

export { howToStepDirectionResolver as a, howToStepResolver as h, index as i };
