// valaxy/dist/node/index.mjs
export { C as ALL_ROUTE, E as EXCERPT_SEPARATOR, G as GLOBAL_STATE, P as PATHNAME_PROTOCOL_RE, V as ViteValaxyPlugins, b as build, c as cli, N as createServer, L as createValaxyPlugin, D as customElements, j as defaultSiteConfig, w as defaultValaxyConfig, F as defaultViteConfig, h as defineAddon, y as defineConfig, k as defineSiteConfig, u as defineTheme, f as defineValaxyAddon, x as defineValaxyConfig, t as defineValaxyTheme, O as encryptContent, g as generateClientRedirects, Q as getGitTimestamp, e as getIndexHtml, M as getServerInfoText, R as isExternal, U as isInstalledGlobally, S as isPath, v as loadConfigFromFile, A as mergeValaxyConfig, m as mergeViteConfigs, p as postProcessForSSG, I as processValaxyOptions, d as registerDevCommand, i as resolveAddonsConfig, Y as resolveImportPath, W as resolveImportUrl, J as resolveOptions, n as resolveSiteConfig, l as resolveSiteConfigFromRoot, o as resolveThemeConfigFromRoot, K as resolveThemeValaxyConfig, q as resolveUserThemeConfig, B as resolveValaxyConfig, z as resolveValaxyConfigFromRoot, r as run, s as ssgBuild, a as startValaxyDev, X as toAtFS, T as transformObject, H as version } from '../shared/valaxy.CEmGx65r.mjs';
import 'node:path';
import 'fs-extra';
import 'consola/utils';
import 'node:process';
import 'define-config-ts';
import 'yargs';
import 'yargs/helpers';
import 'node:os';
import 'consola';
import 'fast-glob';
import 'gray-matter';
import 'js-yaml';
import '@antfu/utils';
import 'debug';
import 'pathe';
import 'defu';
import 'vite';
import 'node:crypto';
import 'floating-vue';
import 'ora';
import 'cross-spawn';
import 'mlly';
import 'resolve-global';
import '@valaxyjs/utils';
import 'node:fs/promises';
import 'dayjs';
import 'feed';
import 'markdown-it';
import 'table';
import 'hookable';
import 'vite-ssg-sitemap';
import 'vite-ssg/node';
import '@intlify/unplugin-vue-i18n/vite';
import '@unhead/addons/vite';
import 'unplugin-vue-components/vite';
import 'vite-plugin-vue-layouts';
import 'vitepress-plugin-group-icons';
import 'markdown-it-async';
import '@shikijs/transformers';
import 'shiki';
import 'css-i18n';
import 'markdown-it-anchor';
import 'markdown-it-attrs';
import 'markdown-it-emoji';
import 'markdown-it-footnote';
import 'markdown-it-image-figures';
import 'markdown-it-task-lists';
import 'node:url';
import 'markdown-it-container';
import 'katex';
import 'unplugin-vue-markdown/vite';
import 'js-base64';
import 'node:fs';
import 'jiti';
import 'unocss';
import 'pascalcase';
import 'lru-cache';
import 'html-to-text';
import 'unplugin-vue-router/vite';
import '@clack/prompts';
import 'node:net';
import 'node:child_process';
import 'node:readline';
import 'qrcode';
import 'ejs';

function $t(key) {
  return `$locale:${key}`;
}

function defineUnoSetup(fn) {
  return fn;
}

export { $t, defineUnoSetup };
