// @unhead/schema-org/dist/index.mjs
export { P as PluginSchemaOrg, S as SchemaOrgUnheadPlugin, U as UnheadSchemaOrg, c as createSchemaOrgGraph, f as defineAddress, g as defineAggregateOffer, h as defineAggregateRating, i as defineArticle, j as defineBook, k as defineBookEdition, l as defineBreadcrumb, n as defineComment, o as defineCourse, p as defineDataset, q as defineEvent, s as defineFoodEstablishment, t as defineHowTo, u as defineHowToStep, v as defineImage, w as defineItemList, x as defineJobPosting, y as defineListItem, z as defineLocalBusiness, A as defineMovie, B as defineMusicAlbum, C as defineMusicGroup, D as defineMusicPlaylist, E as defineMusicRecording, F as defineOffer, G as defineOpeningHours, H as defineOrganization, I as definePerson, J as definePlace, K as definePodcastEpisode, L as definePodcastSeason, M as definePodcastSeries, N as defineProduct, O as defineQuestion, Q as defineReadAction, R as defineRecipe, T as defineReview, d as defineSchemaOrgResolver, V as defineSearchAction, W as defineService, X as defineSoftwareApp, Y as defineTVEpisode, Z as defineTVSeason, _ as defineTVSeries, $ as defineVideo, a0 as defineVirtualLocation, a1 as defineWebPage, a2 as defineWebSite, a5 as imageResolver, m as merge, a3 as normalizeSchemaOrgInput, r as resolveMeta, a as resolveNode, b as resolveNodeId, e as resolveRelation, a4 as useSchemaOrg } from './shared/schema-org.Cp6bpwL2.mjs';
export { aggregateOfferResolver } from './chunks/index.mjs';
export { aggregateRatingResolver } from './chunks/index2.mjs';
export { P as PrimaryArticleId, b as PrimaryBreadcrumbId, d as PrimaryWebPageId, e as PrimaryWebSiteId, a as articleResolver, c as breadcrumbResolver, o as organizationResolver, p as personResolver, w as webPageResolver, f as webSiteResolver } from './chunks/index3.mjs';
export { PrimaryBookId, bookEditionResolver, bookResolver } from './chunks/index4.mjs';
export { commentResolver } from './chunks/index5.mjs';
export { courseResolver } from './chunks/index6.mjs';
export { PrimaryDatasetId, datasetResolver } from './chunks/index7.mjs';
export { placeResolver } from './chunks/index8.mjs';
export { virtualLocationResolver } from './chunks/index9.mjs';
export { PrimaryEventId, eventResolver } from './chunks/index10.mjs';
export { foodEstablishmentResolver } from './chunks/index11.mjs';
export { a as howToStepDirectionResolver, h as howToStepResolver } from './chunks/index12.mjs';
export { HowToId, howToResolver } from './chunks/index13.mjs';
export { itemListResolver } from './chunks/index14.mjs';
export { jobPostingResolver } from './chunks/index15.mjs';
export { listItemResolver } from './chunks/index16.mjs';
export { localBusinessResolver } from './chunks/index17.mjs';
export { movieResolver } from './chunks/index18.mjs';
export { musicAlbumResolver } from './chunks/index19.mjs';
export { musicGroupResolver } from './chunks/index20.mjs';
export { musicPlaylistResolver } from './chunks/index21.mjs';
export { musicRecordingResolver } from './chunks/index22.mjs';
export { offerResolver } from './chunks/index23.mjs';
export { openingHoursResolver } from './chunks/index24.mjs';
export { podcastEpisodeResolver } from './chunks/index25.mjs';
export { podcastSeasonResolver } from './chunks/index26.mjs';
export { podcastSeriesResolver } from './chunks/index27.mjs';
export { addressResolver } from './chunks/index28.mjs';
export { ProductId, productResolver } from './chunks/index29.mjs';
export { questionResolver } from './chunks/index30.mjs';
export { r as ratingResolver } from './shared/schema-org.40UJYB57.mjs';
export { RecipeId, recipeResolver } from './chunks/index31.mjs';
export { reviewResolver } from './chunks/index32.mjs';
export { ServiceId, serviceResolver } from './chunks/index33.mjs';
export { softwareAppResolver } from './chunks/index34.mjs';
export { tvEpisodeResolver } from './chunks/index35.mjs';
export { tvSeasonResolver } from './chunks/index36.mjs';
export { tvSeriesResolver } from './chunks/index37.mjs';
export { videoResolver } from './chunks/index38.mjs';
export { readActionResolver } from './chunks/index39.mjs';
export { searchActionResolver } from './chunks/index40.mjs';
import 'unhead/plugins';
import 'unhead/utils';
import 'ufo';
import './shared/schema-org.Dg61qWpa.mjs';
