// @unhead/schema-org/dist/svelte.mjs
import { useUnhead, useHead } from '@unhead/svelte';
import { s as schemaAutoImports } from './shared/schema-org.CHbRCiep.mjs';
import { U as UnheadSchemaOrg, a3 as normalizeSchemaOrgInput } from './shared/schema-org.Cp6bpwL2.mjs';
export { i as defineArticle, j as defineBook, l as defineBreadcrumb, n as defineComment, o as defineCourse, p as defineDataset, q as defineEvent, s as defineFoodEstablishment, t as defineHowTo, v as defineImage, w as defineItemList, x as defineJobPosting, z as defineLocalBusiness, A as defineMovie, B as defineMusicAlbum, C as defineMusicGroup, D as defineMusicPlaylist, E as defineMusicRecording, H as defineOrganization, I as definePerson, K as definePodcastEpisode, L as definePodcastSeason, M as definePodcastSeries, N as defineProduct, O as defineQuestion, R as defineRecipe, T as defineReview, W as defineService, X as defineSoftwareApp, Y as defineTVEpisode, Z as defineTVSeason, _ as defineTVSeries, $ as defineVideo, a1 as defineWebPage, a2 as defineWebSite } from './shared/schema-org.Cp6bpwL2.mjs';
import 'unhead/plugins';
import 'unhead/utils';
import 'ufo';

function useSchemaOrg(input = [], options = {}) {
  const unhead = options.head || useUnhead();
  unhead.use(UnheadSchemaOrg());
  const entry = useHead(normalizeSchemaOrgInput(input), options);
  const corePatch = entry.patch;
  entry.patch = (input2) => corePatch(normalizeSchemaOrgInput(input2));
  return entry;
}
const schemaOrgAutoImports = [
  {
    from: "@unhead/schema-org/svelte",
    imports: schemaAutoImports
  }
];

export { schemaOrgAutoImports, useSchemaOrg };
